DynamicQuantities support

icweaver opened the pull request:
<!--
* Any PR that is not ready for review should be created as a draft PR.
* Please don't force push to PRs, it removes the history, creates bad notifications, and we will squash and merge in the end anyways.
* Feel free to ping for a review when it passes all tests after a few days (@simondanisch, @ffreyer). We can't guarantee a review in a certain time frame, but we can guarantee to forget about PRs after a while.
* Allowing write access on the PR makes things more convenient, since we can make edits to the PR directly and update it if it gets out of sync with `master` (should be automatic if not disabled).
* Please understand, that some PRs will take very long to get merged. You can do a few things to optimize the time to get it merged:
    * The more tests you add, the easier it is to see that your change works without putting the work on us.
    * The clearer the problem being solved the easier. A PR best only addresses one bug fix or feature.
    * The more you explain the motivation or describe your feature (best with pictures), the easier it will be for us to priorize the PR.
    * Changes with more ambigious benefits are best discussed in a github [discussion](https://github.com/MakieOrg/Makie.jl/discussions) before a PR.
* What deserves a unit test or a reference image test isn't easy to decide, but here are a few pointers:
   * Makie unit tests are preferable, since they're fast to execute and easy to maintain, so if you can add tests to `Makie/src/test`
   * For new recipes or any changes that may get rendered differently by the backends, one should add a reference image test to the best fitting file in `Makie\ReferenceTests\src\tests\**` looking like this:
   ```julia 
   @reference_test "name of test" begin
        # code of test
        ...
        # make sure the last line is a Figure, FigureAxisPlot or Scene
    end
    ``` 
    Adding a reference image test will let your PR fail with the status "n reference images missing", which a maintainer will need to approve and fix. 
    Ideally, a comment with a screenshot of the expected output of the reference image test should be added to the PR.
    We prefer one reference image test with many subplots over multiple reference image tests.
-->
# Description

Fixes #4859

Adds support for DynamicQuantities, almost as a drop-in replacement for Unitful. The PR originally lived here: https://github.com/SymbolicML/DynamicQuantities.jl/pull/165

## Type of change

Delete options that do not apply:

- [x] New feature (non-breaking change which adds functionality)

## Demo

```julia
fig = Figure()

scatter(fig[1, 1], (1:10) .* us"J/s")
scatter(fig[1, 2], (1:10) .* u"K", exp.(1:10) .* us"mm/m^2")
scatter(fig[2, 1], 10 .^ (1:6) .* us"W/m^2", (1:6) .* 1000 .* u"nm"; axis=(; dim2_conversion=Makie.DQConversion(us"μm")))
scatter(fig[2, 2], (0:10) .* u"W/m^2", (0:10) .* u"g"; axis=(; dim1_conversion=Makie.DQConversion(us"W/m^2")))
scatter!(fig[2, 2], (0:10) .* u"kW/m^2", (0:10) .* us"kg")

fig
```
![image](https://github.com/user-attachments/assets/72b24696-df71-487c-bf90-9fe56b13293e)

## Checklist

- [x] Added an entry in CHANGELOG.md (for new features and breaking changes)
- [x] Added or changed relevant sections in the documentation
- [x] Added unit tests for new algorithms, conversion methods, etc.
- [x] Added reference image tests for new plotting functions, recipes, visual options, etc.

cgarling commented on Makie/Project.toml:
```diff
@@ -79,6 +85,7 @@ DelaunayTriangulation = "1.0"
 Distributions = "0.17, 0.18, 0.19, 0.20, 0.21, 0.22, 0.23, 0.24, 0.25"
 DocStringExtensions = "0.8, 0.9"
 Downloads = "1, 1.6"
+DynamicQuantities = "1.8.0"
```
Can this be "1" or "1.8"?

icweaver replied:
For sure, how's this? 08d8ac2

asinghvi17 commented:
Looks like DQ may also have to be added to the extras and test targets of the backend libraries.

icweaver commented:
Thanks all! I have very little experience with the testing machinery in Makie, so apologies in advance for any oversights on my part

Ah, ok, just added some `[extras]` and `[target]` entries (1c007fc). Out of curiosity, is this needed because the backends support down to Julia v1.0, which predates `test/Project.toml`?

icweaver commented:
Ok, I cleaned up the docs a bit. I couldn't seem to import DynamicQuantities in its own scoped `@figure` block, so I just used a `@setup` block for now. Besides that, I think these are the remaining bits that could use some review:

* I made the design decision to not support automatic units, instead preferring explicitness over implicitness. Some [related discussion about this here](https://github.com/MakieOrg/Makie.jl/pull/4940#issuecomment-2874909773). Is this ok with folks?
* AFAICT, DynamicQuantities.jl does not support integration with Dates.jl. Is this a blocker? I've added a note about this [in the docs](https://github.com/icweaver/Makie.jl/blob/fd1ac7c85096a084de81c6c5305ce5c3db4285ca/docs/src/explanations/dim-converts.md?plain=1#L71) for now. (cc: @MilesCranmer)

asinghvi17 commented:
Do you have a link to the doc error @icweaver?  I tried to find it but a brief skim did not reveal any errors

icweaver commented:
Sorry, it was buried with my other commits changing things in the docs. I just split it out in a separate commit so you can take a look (e4af013, error starts on [L9685 in the docs CI log](https://github.com/MakieOrg/Makie.jl/actions/runs/16207894953/job/45762183568?pr=5137#step:7:9685))

So pretty much:

````julia
```@setup DQ_dimconverts
using DynamicQuantities
```

```@figure DQ_dimconverts
using DynamicQuantities

f, ax, pl = scatter((100:100:1_000)us"cm")
```
````

Gives the intended result in the docs:

```julia
using DynamicQuantities

f, ax, pl = scatter((100:100:1_000)us"cm")
```

while trying to combine everything in a fig block seems to throw that error:

````julia
```@figure DQ_dimconverts
using DynamicQuantities

f, ax
```
````

<details><summary>Details</summary>

````julia
┌ Error: failed to run `@example` block in src/explanations/dim-converts.md
  │ ```@example DQ_dimconverts
  │ using CairoMakie
  │ CairoMakie.activate!(; px_per_unit = 2) # hide
  │ import ..MakieDocsHelpers # hide
  │ var"#result" = begin # hide
  │ using DynamicQuantities
  │ 
  │ # Implicitly define figure, axis, and units
  │ f, ax, pl = scatter((100:100:1_000)us"cm")
  │ scatter!(ax, (10:-1:1)u"m")
  │ 
  │ # Implicitly define axis and units
  │ scatter(f[1, 2], (100:100:1_000)us"cm")
  │ scatter!(f[1, 2], (10:-1:1)u"m")
  │ 
  │ # Explicitly define axis, implicitly define units
  │ ax3 = Axis(f[2, 1])
  │ scatter!(ax3, (100:100:1_000)us"cm") # The first plot sets the units
  │ scatter!(ax3, (10:-1:1)u"m")
  │ 
  │ # Explicitly define axis and units
  │ ax4 = Axis(f[2, 2]; dim2_conversion=Makie.DQConversion(us"cm"))
  │ scatter!(ax4, (100:100:1_000)u"cm") # No need for us"cm" now
  │ scatter!(ax4, (10:-1:1)u"m")
  │ 
  │ f
  │ end # hide
  │ MakieDocsHelpers.register_figure!("src/explanations/dim-converts.md", "Dimension conversions", "3d00522", var"#result") # hide
  │ save("3d00522.png", var"#result") # hide
  │ nothing # hide
  │ 
  │ ```
  │   exception =
  │    LoadError: UndefVarError: `@us_str` not defined in `Main.__atexample__named__DQ_dimconverts`
  │    Suggestion: check for spelling errors or missing imports.
  │    Stacktrace:
  │      [1] top-level scope
  │        @ :0
  │      [2] eval
  │        @ ./boot.jl:430 [inlined]
  │      [3] #59
  │        @ ~/.julia/packages/Documenter/eoWm2/src/expander_pipeline.jl:856 [inlined]
  │      [4] cd(f::Documenter.var"#59#61"{Module, Expr}, dir::String)
  │        @ Base.Filesystem ./file.jl:112
  │      [5] (::Documenter.var"#58#60"{Documenter.Page, Module, Expr})()
  │        @ Documenter ~/.julia/packages/Documenter/eoWm2/src/expander_pipeline.jl:855
  │      [6] (::IOCapture.var"#5#9"{DataType, Documenter.var"#58#60"{Documenter.Page, Module, Expr}, IOContext{Base.PipeEndpoint}, IOContext{Base.PipeEndpoint}, IOContext{Base.PipeEndpoint}, IOContext{Base.PipeEndpoint}})()
  │        @ IOCapture ~/.julia/packages/IOCapture/Y5rEA/src/IOCapture.jl:170
  │      [7] with_logstate(f::IOCapture.var"#5#9"{DataType, Documenter.var"#58#60"{Documenter.Page, Module, Expr}, IOContext{Base.PipeEndpoint}, IOContext{Base.PipeEndpoint}, IOContext{Base.PipeEndpoint}, IOContext{Base.PipeEndpoint}}, logstate::Base.CoreLogging.LogState)
  │        @ Base.CoreLogging ./logging/logging.jl:524
  │      [8] with_logger(f::Function, logger::Base.CoreLogging.ConsoleLogger)
  │        @ Base.CoreLogging ./logging/logging.jl:635
  │      [9] capture(f::Documenter.var"#58#60"{Documenter.Page, Module, Expr}; rethrow::Type, color::Bool, passthrough::Bool, capture_buffer::IOBuffer, io_context::Vector{Any})
  │        @ IOCapture ~/.julia/packages/IOCapture/Y5rEA/src/IOCapture.jl:167
  │     [10] runner(::Type{Documenter.Expanders.ExampleBlocks}, node::MarkdownAST.Node{Nothing}, page::Documenter.Page, doc::Documenter.Document)
  │        @ Documenter ~/.julia/packages/Documenter/eoWm2/src/expander_pipeline.jl:854
  │    in expression starting at dim-converts.md:8
  └ @ Documenter ~/.julia/packages/Documenter/eoWm2/src/utilities/utilities.jl:47
  ERROR: LoadError: AssertionError: last_png.id == id
  Stacktrace:
    [1] runner(::Type{FigureBlocks}, node::MarkdownAST.Node{Nothing}, page::Documenter.Page, doc::Documenter.Document)
      @ Main ~/work/Makie.jl/Makie.jl/docs/figure_block.jl:184
    [2] dispatch(::Type{Documenter.Expanders.ExpanderPipeline}, ::MarkdownAST.Node{Nothing}, ::Vararg{Any})
      @ Documenter.Selectors ~/.julia/packages/Documenter/eoWm2/src/utilities/Selectors.jl:170
    [3] expand(doc::Documenter.Document)
      @ Documenter ~/.julia/packages/Documenter/eoWm2/src/expander_pipeline.jl:59
    [4] runner(::Type{Documenter.Builder.ExpandTemplates}, doc::Documenter.Document)
      @ Documenter ~/.julia/packages/Documenter/eoWm2/src/builder_pipeline.jl:224
    [5] dispatch(::Type{Documenter.Builder.DocumentPipeline}, x::Documenter.Document)
      @ Documenter.Selectors ~/.julia/packages/Documenter/eoWm2/src/utilities/Selectors.jl:170
    [6] #88
      @ ~/.julia/packages/Documenter/eoWm2/src/makedocs.jl:280 [inlined]
    [7] withenv(::Documenter.var"#88#90"{Documenter.Document}, ::Pair{String, Nothing}, ::Vararg{Pair{String, Nothing}})
      @ Base ./env.jl:265
    [8] #87
      @ ~/.julia/packages/Documenter/eoWm2/src/makedocs.jl:279 [inlined]
    [9] cd(f::Documenter.var"#87#89"{Documenter.Document}, dir::String)
      @ Base.Filesystem ./file.jl:112
   [10] makedocs(; debug::Bool, format::MarkdownVitepress, kwargs::@Kwargs{sitename::String, pages::Vector{Pair{String, Any}}, expandfirst::Vector{String}, warnonly::Bool, pagesonly::Bool})
      @ Documenter ~/.julia/packages/Documenter/eoWm2/src/makedocs.jl:278
   [11] make_docs(; pages::Vector{Pair{String, Any}})
      @ Main ~/work/Makie.jl/Makie.jl/docs/makedocs.jl:193
   [12] top-level scope
      @ ~/work/Makie.jl/Makie.jl/docs/makedocs.jl:209
  in expression starting at /home/runner/work/Makie.jl/Makie.jl/docs/makedocs.jl:209
````

</details>

icweaver commented:
At any rate, is there anything else in this PR that would be good for me to look into?

SimonDanisch commented on GLMakie/Project.toml:
```diff
@@ -43,3 +43,9 @@ julia = "1"
 path = "../Makie"
+
+[extras]
+DynamicQuantities = "06fc5a27-2a28-4c7c-a15d-362465fb6821"
```
I think this should go into ReferenceTests and not into each backend individually. We moved most test dependencies there, so that backends dont need to worry about it independently.

icweaver replied:
Oh cool, thanks, just added!

icweaver commented:
Just to get things working again, I went back to using the `@setup` block for the documentation. Is there anything else that folks would like me to take a look at?

SimonDanisch commented:
This is a great PR - thank you for your patience and pushing this through!
I've just updated the test images and I hope we can merge it when tests are green.

icweaver commented on CairoMakie/test/Project.toml:
```diff
@@ -1,4 +1,5 @@
 [deps]
+DynamicQuantities = "06fc5a27-2a28-4c7c-a15d-362465fb6821"
```
Ok, just dropped 👍🏾

icweaver replied:
I seem to be running [into this again](https://github.com/MakieOrg/Makie.jl/actions/runs/17057164076/job/48356868112?pr=5137#step:6:411) after removing the backend test deps, e.g.,

```julia
LoadError: ArgumentError: Package DynamicQuantities not found in current path.
  - Run `import Pkg; Pkg.add("DynamicQuantities")` to install the DynamicQuantities package.
```

Is there something different that I should be trying from my end? I'm not sure why the backend tests are not picking up the DynamicQuantities dep from the ReferenceTests source. Sorry if I am missing something obvious!

icweaver replied:
Howdy, just hoping to check back in about this. Thanks!

icweaver replied:
Oh snap, sorry for the oversight, thanks!

SimonDanisch commented:
Thank you for getting this through :)
